File: Sources/SwctxCore/Store.swift
import Foundation
import GRDB
import CryptoKit

/// One SQLite database per indexed workspace, stored under ~/.swctx/indexes/<key>/.
public final class Store: @unchecked Sendable {
    public let pool: DatabasePool
    public let workspaceRoot: URL
    public let workspaceKey: String

    public static let schemaVersion = 2

    public static func key(for root: URL) -> String {
        let digest = SHA256.hash(data: Data(root.standardizedFileURL.path.utf8))
        return digest.prefix(6).map { String(format: "%02x", $0) }.joined()
    }

    public static func baseDir() -> URL {
        let home = FileManager.default.homeDirectoryForCurrentUser
        return home.appendingPathComponent(".swctx", isDirectory: true)
    }

    public static func indexURL(forKey key: String) -> URL {
        baseDir().appendingPathComponent("indexes/\(key)/index.db")
    }

    /// Open (creating if needed) the index DB for a workspace root.
    /// The root is symlink-resolved so discovery and queries share one identity.
    public init(workspaceRoot: URL) throws {
        let root = workspaceRoot.resolvingSymlinksInPath()
        self.workspaceRoot = root
        self.workspaceKey = Store.key(for: root)
        let dir = Store.indexURL(forKey: workspaceKey).deletingLastPathComponent()
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        var config = Configuration()
        config.busyMode = .timeout(10)
        config.prepareDatabase { db in
            try db.execute(sql: "PRAGMA foreign_keys = ON")
        }
        pool = try DatabasePool(path: Store.indexURL(forKey: workspaceKey).path, configuration: config)
        try migrate()
    }

    private func migrate() throws {
        // Skip the write entirely when the schema is already current, so plain
        // queries never open a write transaction on an existing index.
        let current = try? pool.read { db in
            try String.fetchOne(db, sql: "SELECT value FROM meta WHERE key = 'schema_version'")
        }
        if current == String(Store.schemaVersion) { return }
        var migrator = DatabaseMigrator()
        migrator.registerMigration("v1") { db in
            try db.execute(sql: """
                CREATE TABLE meta(key TEXT PRIMARY KEY, value TEXT NOT NULL);
                CREATE TABLE files(
                    id INTEGER PRIMARY KEY,
                    path TEXT NOT NULL UNIQUE,
                    lang TEXT NOT NULL,
                    sha TEXT NOT NULL,
                    size INTEGER NOT NULL,
                    mtime REAL NOT NULL,
                    indexed_at REAL NOT NULL);
                CREATE TABLE chunks(
                    id INTEGER PRIMARY KEY,
                    file_id INTEGER NOT NULL REFERENCES files(id) ON DELETE CASCADE,
                    idx INTEGER NOT NULL,
                    start_line INTEGER NOT NULL,
                    end_line INTEGER NOT NULL,
                    kind TEXT,
                    symbol TEXT,
                    content TEXT NOT NULL,
                    tokens INTEGER NOT NULL DEFAULT 0);
                CREATE INDEX idx_chunks_file ON chunks(file_id);
                CREATE TABLE symbols(
                    id INTEGER PRIMARY KEY,
                    file_id INTEGER NOT NULL REFERENCES files(id) ON DELETE CASCADE,
                    chunk_id INTEGER REFERENCES chunks(id) ON DELETE SET NULL,
                    name TEXT NOT NULL,
                    kind TEXT,
                    line INTEGER NOT NULL,
                    signature TEXT);
                CREATE INDEX idx_symbols_name ON symbols(name);
                CREATE TABLE edges(
                    id INTEGER PRIMARY KEY,
                    src_chunk INTEGER NOT NULL REFERENCES chunks(id) ON DELETE CASCADE,
                    dst_chunk INTEGER REFERENCES chunks(id) ON DELETE SET NULL,
                    dst_name TEXT NOT NULL,
                    kind TEXT NOT NULL,
                    line INTEGER NOT NULL);
                CREATE INDEX idx_edges_src ON edges(src_chunk);
                CREATE INDEX idx_edges_dst ON edges(dst_chunk);
                CREATE INDEX idx_edges_name ON edges(dst_name);
                CREATE TABLE embeddings(
                    chunk_id INTEGER PRIMARY KEY REFERENCES chunks(id) ON DELETE CASCADE,
                    dim INTEGER NOT NULL,
                    vec BLOB NOT NULL);
                """)
            try db.create(virtualTable: "chunks_fts", using: FTS5()) { t in
                t.column("content")
                t.column("symbol")
                t.column("path")
            }
        }
        migrator.registerMigration("v2") { db in
            try db.execute(sql: """
                CREATE TABLE records(
                    id INTEGER PRIMARY KEY,
                    kind TEXT NOT NULL,
                    source TEXT NOT NULL,
                    status TEXT NOT NULL DEFAULT 'completed',
                    title TEXT NOT NULL,
                    payload TEXT NOT NULL,
                    created_at REAL NOT NULL);
                """)
            // External-content FTS over records(title, payload); GRDB's
            // synchronize() emits content='records' plus __records_fts_ai/ad/au
            // triggers, so writes to records keep the index in sync.
            try db.create(virtualTable: "records_fts", using: FTS5()) { t in
                t.column("title")
                t.column("payload")
                t.synchronize(withTable: "records")
            }
            // Receiver/module qualifier for call edges (`m` in `m.f()`);
            // NULL = unqualified bare call. Column exists ahead of the
            // Analyzer EdgeDraft.qualifier field that will populate it.
            try db.execute(sql: "ALTER TABLE edges ADD COLUMN qualifier TEXT")
        }
        try migrator.migrate(pool)
        try pool.write { db in
            try db.execute(
                sql: "INSERT OR REPLACE INTO meta(key, value) VALUES('schema_version', ?), ('workspace_root', ?), ('created_at', ?)",
                arguments: [String(Store.schemaVersion), workspaceRoot.path, String(Date().timeIntervalSince1970)])
        }
    }

    // MARK: - Records

    /// Insert one record and return its row id. The payload is JSON-encoded;
    /// the records_fts sync trigger indexes title + payload automatically.
    @discardableResult
    public func insertRecord(kind: String, source: String, title: String,
                             payload: [String: Any]) throws -> Int64 {
        let data = try JSONSerialization.data(withJSONObject: payload)
        let json = String(decoding: data, as: UTF8.self)
        return try pool.write { db in
            try db.execute(sql: """
                INSERT INTO records(kind, source, title, payload, created_at)
                VALUES(?,?,?,?,?)
                """, arguments: [kind, source, title, json,
                                 Date().timeIntervalSince1970])
            let id = db.lastInsertedRowID
            // Bound the ledger: keep the newest 1000 rows per workspace.
            try db.execute(sql: """
                DELETE FROM records WHERE id NOT IN (
                    SELECT id FROM records ORDER BY id DESC LIMIT 1000)
                """)
            return id
        }
    }

    // MARK: - Registry (~/.swctx/workspaces.json)

    public struct WorkspaceEntry: Codable, Sendable {
        public var path: String
        public var key: String
        public var lastIndexedAt: Double
    }

    public static func registryURL() -> URL {
        baseDir().appendingPathComponent("workspaces.json")
    }

    public static func loadRegistry() -> [WorkspaceEntry] {
        guard let data = try? Data(contentsOf: registryURL()),
              let entries = try? JSONDecoder().decode([WorkspaceEntry].self, from: data)
        else { return [] }
        return entries
    }

    public static func saveRegistry(_ entries: [WorkspaceEntry]) {
        let dir = baseDir()
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        if let data = try? JSONEncoder().encode(entries) {
            try? data.write(to: registryURL(), options: .atomic)
        }
    }

    public static func register(root: URL) {
        var entries = loadRegistry()
        let key = Store.key(for: root)
        if let i = entries.firstIndex(where: { $0.key == key }) {
            entries[i].lastIndexedAt = Date().timeIntervalSince1970
        } else {
            entries.append(WorkspaceEntry(path: root.path, key: key,
                                          lastIndexedAt: Date().timeIntervalSince1970))
        }
        saveRegistry(entries)
    }
}

File: swctx-py/swctx_py/store.py
"""SQLite store — per-workspace index under ~/.swctx-py/indexes/<key>/index.db."""
from __future__ import annotations

import hashlib
import json
import os
import sqlite3
import time
from pathlib import Path

HOME = Path.home() / ".swctx-py"
INDEXES = HOME / "indexes"
MODELS = HOME / "models"
CATALOG = HOME / "workspaces.json"

SCHEMA = """
CREATE TABLE IF NOT EXISTS meta (k TEXT PRIMARY KEY, v TEXT);
CREATE TABLE IF NOT EXISTS files (
    path TEXT PRIMARY KEY, sha TEXT, mtime REAL, lang TEXT);
CREATE TABLE IF NOT EXISTS chunks (
    id INTEGER PRIMARY KEY, file_id TEXT, start_line INT, end_line INT,
    symbol_name TEXT, symbol_type TEXT, content TEXT,
    path_tokens TEXT, symbol_tokens TEXT);
CREATE TABLE IF NOT EXISTS symbols (
    name TEXT, type TEXT, chunk_id INT, file_id TEXT, line INT);
CREATE INDEX IF NOT EXISTS idx_symbols_name ON symbols(name);
CREATE INDEX IF NOT EXISTS idx_chunks_file ON chunks(file_id);
CREATE TABLE IF NOT EXISTS edges (
    id INTEGER PRIMARY KEY AUTOINCREMENT, src_chunk INT, dst_chunk INT,
    dst_name TEXT, kind TEXT, line INT);
CREATE INDEX IF NOT EXISTS idx_edges_src ON edges(src_chunk);
CREATE INDEX IF NOT EXISTS idx_edges_name ON edges(dst_name);
CREATE TABLE IF NOT EXISTS embeddings (
    chunk_id INTEGER PRIMARY KEY, vec BLOB);
CREATE TABLE IF NOT EXISTS records (
    id INTEGER PRIMARY KEY AUTOINCREMENT, kind TEXT, title TEXT,
    body TEXT, created_at REAL);
CREATE TABLE IF NOT EXISTS usage_events (
    id INTEGER PRIMARY KEY AUTOINCREMENT, ts REAL, tool TEXT,
    session TEXT, query TEXT, arg_path TEXT, top_paths TEXT, hit_count INT);
"""


def workspace_key(path: str) -> str:
    return hashlib.sha256(os.path.abspath(path).encode()).hexdigest()[:12]


def index_path(path: str) -> Path:
    return INDEXES / workspace_key(path) / "index.db"


class Store:
    def __init__(self, workspace: str, create: bool = True):
        self.workspace = os.path.abspath(workspace)
        self.key = workspace_key(self.workspace)
        db_path = index_path(self.workspace)
        if not create and not db_path.exists():
            raise FileNotFoundError(f"no index for {self.workspace} — run: swctx-py index {self.workspace}")
        db_path.parent.mkdir(parents=True, exist_ok=True)
        self.db = sqlite3.connect(str(db_path), check_same_thread=False)
        self.db.execute("PRAGMA journal_mode=WAL")
        self.db.executescript(SCHEMA)
        self.fts_ok = self._probe_fts()
        if self.fts_ok:
            self.db.execute(
                "CREATE VIRTUAL TABLE IF NOT EXISTS fts_chunks USING fts5("
                "content, path_tokens, symbol_tokens, folded, chunk_id UNINDEXED)")
        self._vn_corpus: bool | None = None
        self.register()

    def _probe_fts(self) -> bool:
        try:
            self.db.execute("CREATE VIRTUAL TABLE IF NOT EXISTS _fts_probe USING fts5(x)")
            self.db.execute("DROP TABLE IF EXISTS _fts_probe")
            return True
        except sqlite3.OperationalError:
            return False

    def register(self) -> None:
        HOME.mkdir(parents=True, exist_ok=True)
        cat = {}
        if CATALOG.exists():
            try:
                cat = json.loads(CATALOG.read_text())
            except Exception:
                cat = {}
        cat[self.key] = {"path": self.workspace, "updated_at": time.time()}
        CATALOG.write_text(json.dumps(cat, indent=1))

    def meta(self, k: str, default: str | None = None) -> str | None:
        row = self.db.execute("SELECT v FROM meta WHERE k=?", (k,)).fetchone()
        return row[0] if row else default

    def set_meta(self, k: str, v: str) -> None:
        self.db.execute("INSERT OR REPLACE INTO meta VALUES (?,?)", (k, v))
        self.db.commit()

    def log_event(self, tool: str, session: str, query: str = "",
                  arg_path: str = "", top_paths: str = "", hit_count: int = 0) -> None:
        self.db.execute(
            "INSERT INTO usage_events (ts,tool,session,query,arg_path,top_paths,hit_count)"
            " VALUES (?,?,?,?,?,?,?)",
            (time.time(), tool, session, query, arg_path, top_paths, hit_count))
        self.db.commit()

    def corpus_has_vn_filenames(self, path_tokens_of) -> bool:
        """True when ≥1 indexed path carries a VN morpheme (cached)."""
        from .lexicon import VN_MORPHEMES
        if self._vn_corpus is not None:
            return self._vn_corpus
        found = False
        for (p,) in self.db.execute("SELECT path FROM files"):
            if VN_MORPHEMES.intersection(path_tokens_of(p).split()):
                found = True
                break
        self._vn_corpus = found
        return found
